### PR TITLE
`F.tree_lstm` test for ChainerX

### PR DESCRIPTION
**Pull request by dido1998**
The double backward tests have to be skipped due to the following error:
`RuntimeError: cannot twice-differentiate an old style Function "TreeLSTM"`

### CODE REVIEW

File: tests/chainer_tests/functions_tests/rnn_tests/test_function_tree_lstm.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+    def forward(self, inputs, device):
+        out = functions.tree_lstm(*list(inputs))
+        return out
+
+    def forward_expected(self, inputs):
+        def _extract_gates(x, n_split=5):
+            r = x.reshape(
+                (x.shape[0], n_split, x.shape[1] // n_split) + x.shape[2:])
+            return (r[:, i, :] for i in six.moves.range(n_split))
```
Comment (asi1024) [inline tests/chainer_tests/functions_tests/rnn_tests/test_function_tree_lstm.py]: This function seems to return a generator. Why not return a list?

File: tests/chainer_tests/functions_tests/rnn_tests/test_function_tree_lstm.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+            'cudnn_deterministic': [True, False],
+            'autotune': [True, False],
+        })]))
+class TestTreeLSTM(testing.FunctionTestCase):
+
+    dodge_nondifferentiable = True
+
+    def setUp(self):
+        self.check_forward_options.update({
+            'rtol': 1e-2, 'atol': 1e-2})
```
Comment (asi1024) [inline tests/chainer_tests/functions_tests/rnn_tests/test_function_tree_lstm.py]: The tolerances seem too large. How about using large tolerances only in float16 test?

File: tests/chainer_tests/functions_tests/rnn_tests/test_function_tree_lstm.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        })
+        # With cuDNN
+        + testing.product({
+            'use_cudnn': ['always'],
+            'cudnn_deterministic': [True, False],
+            'autotune': [True, False],
+        })]))
+class TestTreeLSTM(testing.FunctionTestCase):
+
+    dodge_nondifferentiable = True
```
Comment (asi1024) [inline tests/chainer_tests/functions_tests/rnn_tests/test_function_tree_lstm.py]: Does `tree_lstm` have a non-differentiable point?

File: tests/chainer_tests/functions_tests/rnn_tests/test_function_tree_lstm.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+    def setUp(self):
+        self.check_forward_options.update({
+            'rtol': 1e-2, 'atol': 1e-2})
+        self.check_backward_options.update({
+            'rtol': 1e-2, 'atol': 1e-2})
+        self.check_double_backward_options.update({
+            'rtol': 5e-3, 'atol': 5e-2})
+        self.skip_double_backward_test = True
```
Comment (asi1024) [inline tests/chainer_tests/functions_tests/rnn_tests/test_function_tree_lstm.py]: Could you explain why double backward tests should be skipped?

Comment (dido1998) [reply tests/chainer_tests/functions_tests/rnn_tests/test_function_tree_lstm.py]: I am getting the following error while running the backward test:
`RuntimeError: cannot twice-differentiate an old style Function "TreeLSTM"`

File: tests/chainer_tests/functions_tests/rnn_tests/test_function_tree_lstm.py (hunk 1 above)
Comment (dido1998) [reply tests/chainer_tests/functions_tests/rnn_tests/test_function_tree_lstm.py]: I have changed it to a list.

File: tests/chainer_tests/functions_tests/rnn_tests/test_function_tree_lstm.py (hunk 2 above)
Comment (dido1998) [reply tests/chainer_tests/functions_tests/rnn_tests/test_function_tree_lstm.py]: Done

File: tests/chainer_tests/functions_tests/rnn_tests/test_function_tree_lstm.py (hunk 3 above)
Comment (dido1998) [reply tests/chainer_tests/functions_tests/rnn_tests/test_function_tree_lstm.py]: Yes, you're right it doesn't. I have removed it.

**Comment by dido1998** [comment]
@asi1024 could you check this?

File: tests/chainer_tests/functions_tests/rnn_tests/test_function_tree_lstm.py (hunk 4 above)
Comment (asi1024) [reply tests/chainer_tests/functions_tests/rnn_tests/test_function_tree_lstm.py]: I see. Could you leave a comment about it in the description of this PR?

**Comment by asi1024** [comment]
pfnCI, test this please.

File: tests/chainer_tests/functions_tests/rnn_tests/test_function_tree_lstm.py (hunk 4 above)
Comment (dido1998) [reply tests/chainer_tests/functions_tests/rnn_tests/test_function_tree_lstm.py]: Done.

**Comment by asi1024** [comment]
Jenkins failures seem unrelated to this PR. LGTM.

**Comment by dido1998** [comment]
can this be merged?

**Comment by asi1024** [comment]
pfnCI, test this please.

**Comment by asi1024** [comment]
LGTM. Thank you for the PR!

**Pull request merged**